actionpack/lib/action_dispatch/routing/polymorphic_routes.rb

```diff
@@ -228,8 +227,12 @@ def self.polymorphic_method(recipient, record_or_hash_or_array, action, type, op
             else
               method, args = builder.handle_model record_or_hash_or_array
             end
+            method_type = method.chomp "_path"
 
-            if options.empty?
+            if method_type == method_type.singularize
+              options[:format] ||= nil
+              recipient.public_send(method, *args, options)
```

Comment:
Thanks for giving some time to write some feedback!

>If this is the right patch I think you don't need to put it within the if .. elsif options.empty block.
```Ruby

method_type = method.chomp("_path")
if method_type == method_type.singularize
  options[:format] ||= nil
end

if options.empty?
  recipient.public_send(method, *args)
else
  recipient.public_send(method, *args, options)
end
```
This is a good idea for sure but I do think we need this because we want to know if the `options` hash is truly empty.  If you look at #43766 they were having issues rendering because we were always passing in `format: nil`.  By keeping it in the else block we still satisfy the conditions of the original fix implemented in #43766.

>Also It might be very overhead for this PR ... but It would be very explainable if we had a method in String that can check if a word is singular?
```Ruby
class String
  def singular?
     raise ArgumentError, "The input contains multiple words" if self.split.size > 1
     self.singularize == word
  end
end
```
I totally agree! Feel free to open a PR for this but I think it might be a little out of scope here.

>Also it might more declarative to use delete_suffix than chomp
In my experience chomp is to remove line separators.

Love this idea! Implementing it now!  



The one liner is very slick but I think I'll keep it as is to keep it readable for future contributors.

Reply:
> This is a good idea for sure but I do think we need this because we want to know if the options hash is truly empty. If you look at https://github.com/rails/rails/pull/43766 they were having issues rendering because we were always passing in format: nil. By keeping it in the else block we still satisfy the conditions of the original fix implemented in https://github.com/rails/rails/pull/43766.

Got it. Also you can use `return` for flow control. 
I think it might work. I run the tests that you made and they've passed. 
Also It's been used in the same file.  But you might already thought about it and discarded it.

```rruby
if options.empty?
  return recipient.public_send(method, *args)
end 

method_type = method.delete_suffix("_path")
options[:format] ||= nil if method_type == method_type.singularize

recipient.public_send(method, *args, options)
```

Reply:
You have great code ideas!  So the main reason I haven't done this is because we need to check for the singular resource first in order to get this fix compliant with #43766. 

I just tested this and it failed:
```Ruby
......................F

Failure:
TestGenerationPrefix::WithMountedEngine#test_[OBJECT]_generating_engine's_route_with_polymorphic_url_from_singular_resource [/Users/paulreece/rails/actionpack/test/dispatch/prefix_generation_test.rb:351]:
Expected: "/awesome/blog/author"
  Actual: "/awesome/blog/author.1"
```

I need to check for a singular resource before calling `recipient.public_send(method, *args)` for empty options.  Otherwise the bug will still render the `.1` on the end because it will early return from the `options.empty?` conditional.  This is why I've put the `singular` check before it in a conditional as well.

Reply:
You did inspire me to change it to this though:
```Ruby
            if options.empty? && method_type != method_type.singularize
              recipient.public_send(method, *args)
            else
              options[:format] ||= nil
              recipient.public_send(method, *args, options)
            end
```